src/crimson/os/seastore/cache.cc

```diff
@@ -31,20 +31,29 @@ Cache::retire_extent_ret Cache::retire_extent_addr(
   Transaction &t, paddr_t addr, extent_len_t length)
 {
   LOG_PREFIX(Cache::retire_extent);
-  if (auto ext = t.write_set.find_offset(addr); ext != t.write_set.end()) {
-    DEBUGT("found {} in t.write_set", t, addr);
+  CachedExtentRef ext;
+  auto result = t.get_extent(addr, &ext);
+  if (result == Transaction::get_extent_ret::PRESENT) {
+    DEBUGT("found {} in t", t, addr);
     t.add_to_retired_set(CachedExtentRef(&*ext));
     return retire_extent_iertr::now();
-  } else if (auto iter = extents.find_offset(addr);
-      iter != extents.end()) {
-    auto ret = CachedExtentRef(&*iter);
+  } else if (result == Transaction::get_extent_ret::RETIRED) {
+    ERRORT("{} is already retired", t, addr);
+    ceph_abort();
+  }
+
+  // absent from transaction
+  result = query_cache_for_extent(addr, &ext);
+  if (result == Transaction::get_extent_ret::PRESENT) {
+    t.add_to_read_set(ext);
     return trans_intr::make_interruptible(
-      ret->wait_io()
-    ).then_interruptible([&t, ret=std::move(ret)]() mutable {
-      t.add_to_retired_set(ret);
+      ext->wait_io()
+    ).then_interruptible([&t, ext=std::move(ext)]() mutable {
+      t.add_to_retired_set(ext);
       return retire_extent_iertr::now();
     });
-  } else {
+  } else { // result == get_extent_ret::ABSENT
+    // FIXME leverage RetiredExtentPlaceholder
```

Comment:
I don't understand this comment.

Reply:
IIUC there is a missing piece of interruptive transaction invalidation here.

Say that we have transaction A and B, conflicting on extent E (at laddr L):
```
1)  A: dec_ref(L)
2)  B: read(L) -> E
3)  B: mutate(E)
4)  B: submit_transaction() // assume successful
// A cannot be invalidated because E is not in A's read-set
.......
```

I think we have two options to solve this:
1. create a `RetiredExtentPlaceholder` E' at `1)` to add E' to both A's `read-set`, A's `retired-set`, and `Cache::extents`. When B tries to read out E at `2)`, `Cache` transparently replace E' by the real E in `extents` before return to B, and synchronously tell A to also replace E' by E. This is possible because E' has back reference to A;
2. simpler: fall back to read the real E out during `1) dec_ref(L)`;

What do you think?

Reply:
Yep, that's right.  Probably worth adding this example and explanation to the FIXME.

We're going to want to avoid 2., btw.  Deleting an object will mean dec_ref'ing all of its constituent extents, and we want to avoid actually reading them.

Reply:
OK, I'll go for option 1.

Improved FIXME to include the example.